lib/Dialect/AIEX/Transforms/AIEDmaToNpu.cpp

```diff
@@ -456,56 +456,99 @@ struct WriteBdToBlockWritePattern : OpConversionPattern<NpuWriteBdOp> {
-    words[2] |= (op.getEnablePacket() & 0x1) << 30;
-    words[2] |= (op.getOutOfOrderId() & 0x3f) << 24;
-    words[2] |= (op.getPacketId() & 0x1f) << 19;
-    words[2] |= (op.getPacketType() & 0x7) << 16;
-
-    // DMA_BDX_3
-    // TODO: Secure Access
-    words[3] |= (op.getD0Size() & 0x3ff) << 20;
-    words[3] |= op.getD0Stride() & 0xfffff;
-
-    // DMA_BDX_4
-    words[4] = 0x80000000; // burst length;
-    words[4] |= (op.getD1Size() & 0x3ff) << 20;
-    words[4] |= op.getD1Stride() & 0xfffff;
-
-    // DMA_BDX_5
-    // TODO: SIMID, AxCache, AXQoS
-    words[5] = op.getD2Stride() & 0xfffff;
-
-    // DMA_BDX_6
-    words[6] |= (op.getIterationCurrent() & 0x3f) << 26;
-    words[6] |= (op.getIterationSize() & 0x3f) << 20;
-    words[6] |= op.getIterationStride() & 0xfffff;
-
-    // DMA_BDX_7
-    // TODO: TLAST Suppress
-    words[7] |= (op.getNextBd() & 0xf) << 27;
-    words[7] |= (op.getUseNextBd() & 0x1) << 26;
-    words[7] |= (op.getValidBd() & 0x1) << 25;
-    words[7] |= (op.getLockRelVal() & 0xef) << 18;
-    words[7] |= (op.getLockRelId() & 0xf) << 13;
-    words[7] |= (op.getLockAcqEnable() & 0x1) << 12;
-    words[7] |= (op.getLockAcqVal() & 0xef) << 5;
-    words[7] |= op.getLockAcqId() & 0xf;
-
+    auto bd_id = op.getBdId();
+    uint32_t bd_addr;
+    if (op.getRow() == 0) {
+      bd_addr = (op.getColumn() << tm.getColumnShift()) |
+                (op.getRow() << tm.getRowShift()) | (0x1D000 + bd_id * 0x20);
+
+      // DMA_BDX_0
+      words[0] = op.getBufferLength();
+
+      // DMA_BDX_1
+      words[1] = op.getBufferOffset();
+
+      // DMA_BDX_2
+      // En Packet , OoO BD ID , Packet ID , Packet Type
+      words[2] |= (op.getEnablePacket() & 0x1) << 30;
+      words[2] |= (op.getOutOfOrderId() & 0x3f) << 24;
+      words[2] |= (op.getPacketId() & 0x1f) << 19;
+      words[2] |= (op.getPacketType() & 0x7) << 16;
+
+      // DMA_BDX_3
+      // TODO: Secure Access
+      words[3] |= (op.getD0Size() & 0x3ff) << 20;
+      words[3] |= op.getD0Stride() & 0xfffff;
+
+      // DMA_BDX_4
+      words[4] = 0x80000000; // burst length;
+      words[4] |= (op.getD1Size() & 0x3ff) << 20;
+      words[4] |= op.getD1Stride() & 0xfffff;
+
+      // DMA_BDX_5
+      // TODO: SIMID, AxCache, AXQoS
+      words[5] = op.getD2Stride() & 0xfffff;
+
+      // DMA_BDX_6
+      words[6] |= (op.getIterationCurrent() & 0x3f) << 26;
+      words[6] |= (op.getIterationSize() & 0x3f) << 20;
+      words[6] |= op.getIterationStride() & 0xfffff;
+
+      // DMA_BDX_7
+      // TODO: TLAST Suppress
+      words[7] |= (op.getNextBd() & 0xf) << 27;
+      words[7] |= (op.getUseNextBd() & 0x1) << 26;
+      words[7] |= (op.getValidBd() & 0x1) << 25;
+      words[7] |= (op.getLockRelVal() & 0xef) << 18;
+      words[7] |= (op.getLockRelId() & 0xf) << 13;
+      words[7] |= (op.getLockAcqEnable() & 0x1) << 12;
+      words[7] |= (op.getLockAcqVal() & 0xef) << 5;
+      words[7] |= op.getLockAcqId() & 0xf;
+    } else {
```

Comment:
What about worker tiles? Also some devices might have more than one row of memtiles. Check against the device model